### ISSUE TITLE
"Can't set headers after they are sent" error when updating model with associations

### ISSUE THREAD

**Issue by chotiwat** [balderdashy/sails#2922]
I tracked down the problem and have found that it is a bug in waterline module. I have already opened a pull request here: balderdashy/waterline#1002.

**Below is how the problem happened:**
I have two models, Pet and User, which are defined as follows:
```js
// api/models/Pet.js
module.exports = {
  attributes: {
    name: 'STRING',
    color: 'STRING'
  }
};

// api/models/User.js
module.exports = {
  attributes: {
    name: 'STRING',
    age: 'INTEGER',
    pony: { model: 'pet' },
    unicorn: { model: 'pet' }
  }
};
```

Then I successfully created a User model:
```http
POST /user HTTP/1.1
Content-Length: 116
Content-Type: application/x-www-form-urlencoded

name=Mike&age=21&pony%5Bname%5D=Pinkie+Pie&pony%5Bcolor%5D=pink&unicorn%5Bname%5D=Holy+Horn&unicorn%5Bcolor%5D=white
```
Response:
```json
{
  "name": "Mike",
  "age": 21,
  "pony": 9,
  "unicorn": 10,
  "createdAt": "2015-05-11T07:38:31.181Z",
  "updatedAt": "2015-05-11T07:38:31.181Z",
  "id": 3
}
```

However, whenever I tried to update `pony` and `unicorn` of the user, the server would crash.
```http
PUT /user/3 HTTP/1.1
Content-Length: 55
Content-Type: application/x-www-form-urlencoded

pony%5Bname%5D=Pinkie+Puff&unicorn%5Bcolor%5D=%23ffffff
```
Server log:
```
silly: Published  user  to  sails_model_user_3:update
silly: Published  user  to  sails_model_user_3:update
silly: res.ok() :: Sending 200 ("OK") response
silly: res.ok() :: Sending 200 ("OK") response
verbose: Lowering sails...
verbose: Sent kill signal to child process (38186)...
verbose: Shutting down socket server...
verbose: Shutting down HTTP server...

http.js:707
    throw new Error('Can\'t set headers after they are sent.');
          ^
Error: Can't set headers after they are sent.
    at ServerResponse.OutgoingMessage.setHeader (http.js:707:11)
    at ServerResponse.res.setHeader (/Users/top/Playground/sails-app/node_modules/sails/node_modules/express/node_modules/connect/lib/patch.js:133:22)
    at ServerResponse.res.set.res.header (/Users/top/Playground/sails-app/node_modules/sails/node_modules/express/lib/response.js:577:10)
    at ServerResponse.res.send (/Users/top/Playground/sails-app/node_modules/sails/node_modules/express/lib/response.js:142:12)
    at ServerResponse.res.json (/Users/top/Playground/sails-app/node_modules/sails/node_modules/express/lib/response.js:223:15)
    at ServerResponse.jsonx (/Users/top/Playground/sails-app/node_modules/sails/lib/hooks/responses/index.js:218:21)
    at Object.sendOK (/Users/top/Playground/sails-app/api/responses/ok.js:28:16)
    at ServerResponse.bound [as ok] (/Users/top/Playground/sails-app/node_modules/sails/node_modules/lodash/dist/lodash.js:729:21)
    at foundAgain (/Users/top/Playground/sails-app/node_modules/sails/lib/hooks/blueprints/actions/update.js:86:13)
    at bound (/Users/top/Playground/sails-app/node_modules/sails/node_modules/lodash/dist/lodash.js:957:21)
    at applyInOriginalCtx (/Users/top/Playground/sails-app/node_modules/sails/node_modules/waterline/lib/waterline/utils/normalize.js:421:80)
    at wrappedCallback (/Users/top/Playground/sails-app/node_modules/sails/node_modules/waterline/lib/waterline/utils/normalize.js:320:18)
    at _normalizeCallback.callback.success (/Users/top/Playground/sails-app/node_modules/sails/node_modules/waterline/node_modules/switchback/lib/normalize.js:33:31)
    at _switch (/Users/top/Playground/sails-app/node_modules/sails/node_modules/waterline/node_modules/switchback/lib/factory.js:48:28)
    at returnResults (/Users/top/Playground/sails-app/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/basic.js:168:9)
    at /Users/top/Playground/sails-app/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/basic.js:79:16
```

**Comment by tjwebb** [comment]
Post your code. This means that you've already send a response to the client earlier in your request handler.

**Comment by chotiwat** [comment]
@tjwebb What do you mean? I already provided you with all the details including my code: those two model definitions. It is a bug in waterline module which I have fixed in balderdashy/waterline#1002. The pull request is now merged. Please see the pull request for more technical details.

**Comment by sgress454** [comment]
Hi @chotiwat, thanks for the report and the PR.  In the future it'd probably be a bit easier for us to debug if you could send your example requests as JSON instead urlencoded!  I can confirm that this is fixed in the new version of Waterline.
